### PR TITLE
Add codec to encode/decode GRPC data frame.

### PR DESCRIPTION
**Pull request by fengli79**
Add codec to encode/decode GRPC data frame.

### CODE REVIEW

**Comment by PiotrSikora** [comment]
Is there any reason why this is written from scratch and not using https://github.com/grpc/grpc?

**Comment by mattklein123** [comment]
Per @PiotrSikora it would be very, very, very nice if we could work with the gRPC c++ folks to refactor slightly so that we can consume the generated RPC stubs (codec, etc.) but provide our own transport. I already replicated some stuff for unary calls out of expediency, but now that you are all involved it would be great to fix this for real. Can we at least SWAG what it would take before reimplementing?

**Comment by louiscryan** [comment]
@ctiller - Hey Craig! Its probably a good time to start discussing some of the integration points between Envoy and GRPC

In this particular instance WDYT? Would it be easy to factor out the framing code from core? If not then this should probably proceed as is and we have a longer term goal around factoring

**Comment by fengli79** [comment]
Any chance to get this submitted? Refactor GRPC c core to reuse the framer is not trivial and won't happen any time soon.
This pull request is to unblock the data path for grpc-web. For the control plane GRPC requests initiated by envoy can be a different topic.

Here's the situation:
1. for grpc-web, we need a generic client, means the target grpc method is denoted by the request on the wire. Thus no code gen needed.
2. for grpc request to control plane services, code gen might be helpful, but not a blocker, if code gen is highly desired, you can start to use it anyway, remained part is merely optimization.

**Comment by mattklein123** [comment]
@fengli79 I wasn't aware this was ready for review. @RomanDzhabarov can you take a first pass?

**Comment by RomanDzhabarov** [comment]
will check this tonight

File: source/common/grpc/codec.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+class Decoder {
+public:
+  Decoder();
+
+  // Decodes the given buffer with GRPC data frame.
+  // @param input supplies the binary octets wrapped in a GRPC data frame.
+  // @param input supplies the buffer to store the decoded data.
+  // @return bool whether the decoding success.
+  bool Decode(Buffer::Instance& input, std::vector<Frame>* output);
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: usually in envoy we pass by ref in this case (less to deal with null checks, alloc/dealloc/etc).

File: source/common/grpc/codec.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  void NewFrame(uint8_t flags, uint64_t length, uint8_t* output);
+};
+
+class Decoder {
+public:
+  Decoder();
+
+  // Decodes the given buffer with GRPC data frame.
+  // @param input supplies the binary octets wrapped in a GRPC data frame.
+  // @param input supplies the buffer to store the decoded data.
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: fix param name

File: source/common/grpc/codec.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+class Encoder {
+public:
+  Encoder();
+
+  // Creates a new GRPC data frame with the given flags and length.
+  // @param flags supplies the GRPC data frame flags.
+  // @param length supplies the GRPC data frame length.
+  // @param output the buffer to store the encoded data, it's size must be 5.
+  void NewFrame(uint8_t flags, uint64_t length, uint8_t* output);
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: using std::array<uint8_t, 5>? (would be good to enforce size 5 requirement here)

File: source/common/grpc/codec.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  Decoder();
+
+  // Decodes the given buffer with GRPC data frame.
+  // @param input supplies the binary octets wrapped in a GRPC data frame.
+  // @param input supplies the buffer to store the decoded data.
+  // @return bool whether the decoding success.
+  bool Decode(Buffer::Instance& input, std::vector<Frame>* output);
+
+private:
+  // Wiring format of GRPC data frame header:
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: can you also put a link to official docs if any for this format?

File: source/common/grpc/codec.h (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: newFrame

File: source/common/grpc/codec.h (hunk 1 above)
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: decode(

File: source/common/grpc/codec.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+const uint8_t GRPC_FH_DEFAULT = 0b0u;
+// Last bit for a compressed message.
+const uint8_t GRPC_FH_COMPRESSED = 0b1u;
+
+enum class CompressionAlgorithm { None, Gzip };
+
+struct Frame {
+  uint8_t flags;
+  uint32_t length;
+  Buffer::Instance* data;
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.h]: Buffer::InstancePtr

Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h (hunk 3 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done

Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h (hunk 2 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h (hunk 1 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h (hunk 4 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include <array>
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: del or add to precompiled header

File: source/common/grpc/codec.h [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+namespace Grpc {
+// Last bit for an expanded message without compression.
+const uint8_t GRPC_FH_DEFAULT = 0b0u;
+// Last bit for a compressed message.
+const uint8_t GRPC_FH_COMPRESSED = 0b1u;
+
+enum class CompressionAlgorithm { None, Gzip };
+
+struct Frame {
+  uint8_t flags;
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: _ postfix for members (.e.g, flags_)

File: source/common/grpc/codec.h [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+class Encoder {
+public:
+  Encoder();
+
+  // Creates a new GRPC data frame with the given flags and length.
+  // @param flags supplies the GRPC data frame flags.
+  // @param length supplies the GRPC data frame length.
+  // @param output the buffer to store the encoded data, it's size must be 5.
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: "... data. Its size must be 5."

File: source/common/grpc/codec.h [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+class Decoder {
+public:
+  Decoder();
+
+  // Decodes the given buffer with GRPC data frame.
+  // @param input supplies the binary octets wrapped in a GRPC data frame.
+  // @param output supplies the buffer to store the decoded data.
+  // @return bool whether the decoding success.
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: "... succeeded or not."

File: source/common/grpc/codec.h [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  bool decode(Buffer::Instance& input, std::vector<Frame>& output);
+
+private:
+  // Wiring format (http://www.grpc.io/docs/guides/wire.html) of GRPC data frame
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: "Wire format ..."

File: source/common/grpc/codec.h [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  // header:
+  //
+  // -----------------------------------------------------------------------
+  // |R|R|R|R|R|R|R|R|C|      L     |      L     |      L     |      L     |
+  // -----------------------------------------------------------------------
+  //    Flag (1 byte)                Message Length (4 bytes)
+  //
+  // A fixed header consists of five bytes.
+  // The first byte is the Flag. The last one "C" bit indicates if the message
+  // is compressed or not (0 is uncompressed, 1 is compressed). The rest seven
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: The other seven

File: source/common/grpc/codec.h [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  // "R" bits is reserved for future use.
```
Comment (mattklein123) [inline source/common/grpc/codec.h]: are reserved

File: test/common/grpc/codec_test.cc [hunk 13]
```diff
@@ -0,0 +1,115 @@
+#include <array>
```
Comment (mattklein123) [inline test/common/grpc/codec_test.cc]: del or add to precompiled header

File: test/common/grpc/codec_test.cc [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+#include "common/buffer/buffer_impl.h"
+#include "common/grpc/codec.h"
+#include "test/generated/helloworld.pb.h"
```
Comment (mattklein123) [inline test/common/grpc/codec_test.cc]: nit: newline before this line

File: test/common/grpc/codec_test.cc [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+  for (Frame& frame : frames) {
+    EXPECT_EQ(GRPC_FH_DEFAULT, frame.flags);
+    EXPECT_EQ(static_cast<uint64_t>(request.ByteSize()), frame.length);
+
+    helloworld::HelloRequest result;
+    result.ParseFromArray(frame.data->linearize(frame.data->length()), frame.data->length());
+    EXPECT_EQ("hello", result.name());
+  }
+}
+} // Grpc
```
Comment (mattklein123) [inline test/common/grpc/codec_test.cc]: nit: newline before this line

File: source/common/grpc/codec.h (hunk 6 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: source/common/grpc/codec.h (hunk 7 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Is it needed for public struct?

File: source/common/grpc/codec.h (hunk 8 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: source/common/grpc/codec.h (hunk 9 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: source/common/grpc/codec.h (hunk 10 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done

File: source/common/grpc/codec.h (hunk 11 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: source/common/grpc/codec.h (hunk 12 above)
Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: test/common/grpc/codec_test.cc (hunk 13 above)
Comment (fengli79) [inline test/common/grpc/codec_test.cc]: Done.

File: test/common/grpc/codec_test.cc (hunk 14 above)
Comment (fengli79) [inline test/common/grpc/codec_test.cc]: Done.

File: test/common/grpc/codec_test.cc (hunk 15 above)
Comment (fengli79) [inline test/common/grpc/codec_test.cc]: Done.

File: source/common/grpc/codec.cc [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+        } else {
+          frame_.data->add(mem, remain_in_frame);
+          mem += remain_in_frame;
+          j += remain_in_frame;
+        }
+        if (frame_.length == frame_.data->length()) {
+          output.push_back(std::move(frame_));
+          frame_.flags = 0;
+          frame_.length = 0;
+          frame_.data.reset();
```
Comment (RomanDzhabarov) [inline source/common/grpc/codec.cc]: do not think there is a need to .reset() here as ownership was transferred to the frame in output

Comment (fengli79) [inline source/common/grpc/codec.cc]: Right, done.

File: source/common/grpc/codec.h (hunk 7 above)
Comment (mattklein123) [inline source/common/grpc/codec.h]: We have been doing this for public structs so to be consistent I would change it. (I don't really care that much about this but might as well be consistent).

Comment (fengli79) [inline source/common/grpc/codec.h]: Done.

File: test/common/grpc/codec_test.cc [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  Buffer::OwnedImpl buffer;
+  std::array<uint8_t, 5> header;
+  Encoder encoder;
+  encoder.newFrame(GRPC_FH_DEFAULT, request.ByteSize(), header);
+  buffer.add(header.data(), 5);
+  buffer.add(request.SerializeAsString());
+
+  std::vector<Frame> frames;
+  Decoder decoder;
+  decoder.decode(buffer, frames);
```
Comment (RomanDzhabarov) [inline test/common/grpc/codec_test.cc]: EXPECT_TRUE and in the case below

**Comment by RomanDzhabarov** [comment]
LGTM, one comment left

File: test/common/grpc/codec_test.cc (hunk 17 above)
Comment (fengli79) [inline test/common/grpc/codec_test.cc]: Done.

File: source/common/grpc/codec.cc [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+          output.push_back(std::move(frame_));
+          frame_.flags_ = 0;
+          frame_.length_ = 0;
+          state_ = State::FH_FLAG;
+        }
+        break;
+      }
+    }
+  }
+  return true;
```
Comment (lizan) [inline source/common/grpc/codec.cc]: We might want to check `state_` to see if the input was a partial GRPC frame? (e.g. frame length = 100 but only 10 bytes in input buffer.
Also add a test for that case.

Comment (fengli79) [inline source/common/grpc/codec.cc]: Test added.
You don't need to check the state_. If grpc call get aborted when decoding, you should be able to know that from the HTTP2 layer, as the stream will be reset. Any intermediate state and buffer can be discarded. The workflow would be keep feeding data to the decoder as they arrive and check if there's one or more complete frame get decoded.

**Comment by mattklein123** [comment]
@fengli79 one other thing came to mind when I read @lizan comment. You aren't draining the buffer anywhere. I'm not sure if this is intentional, but either way I think some changes are needed:
1) If you intend for the caller to drain the buffer, you should change decode() to take a const Buffer::Instance&, and document that decode consumes all bytes if it returns true. Then the caller can drain.
2) Have decode() drain the buffer itself, and document that it will drain all data, and add tests to this effect.

**Comment by fengli79** [comment]
It's not intended, the caller need to drain the data when needed, as the
test did.

**Comment by fengli79** [comment]
The decoder takes all the data, no matter it returns true or false. It can
be written in a way to drain the data always to make it work like a move.
However let the caller drain data as needed is more flexible. The envoy
buffer is a wrapper which is not very convenient for decoding, to get best
performance, I would prefer to get a buffer can be modified in flight, like
remove/modify byte in a slice and add/remove slice in the buffer.

**Comment by mattklein123** [comment]
OK that's fine. Then just make the param const and document that all bytes are consumed if it's not already documented.

**Comment by fengli79** [comment]
@mattklein123 With a second think, I change it to drain the inputs when decoding succeeded. Also change the CodecTest to GrpcCodecTest to avoid naming conflicts with other tests.

**Comment by mattklein123** [comment]
@fengli79 looks good. I forgot that you need to sign CLA. Can you ACK back when done and I will merge.

**Comment by fengli79** [comment]
Signed.

**Pull request merged**